### ISSUE TITLE
Spec should define how the feature interacts with blocking tools (filterlists etc.)

### ISSUE THREAD

**Issue by pes10k** [w3c/webtransport#96]
Blocking tools (uBlock Origin, AdBlock Plus, Disconnect, etc) are important parts of privacy on the web.  The proposal has several aspects that make it not obvious how blocking tools would interact with this functionality.  Partial list:

 - If I'm mulitplexing over the existing HTTP3 connection, how can I block the Web Transport request (since no new request is happening)
 - Most blocking tools focus on blocking the initial request.  The spec seems to envision connections that require no additional request, just receiving server push messages (for example, layering over HTTP3 and just listening).  How can I block these?
 - Can blocking tools see into the communication, and block some messages but not others?

The above is just a partial list, but meant as motivating ambiguous (or, at least to an outsider, unclear) cases.  Having the spec more clearly define how, and with what decision, blocking tools can make decisions, would be very useful in understanding the spec's privacy properties.

**Comment by ricea** [comment]
Extensions are technically not part of the web platform, and so are not normally covered by standards.

I would expect each browser would define their own blocking semantics. My expectation is that blocking would happen immediately after the creation of the QuicTransport or Http3Transport object.

Blocking on a per-message basis would likely incur considerable overhead, and so would be less likely to be implemented.

**Comment by snyderp** [comment]
Generally how this has been done in the past is to define (in normative text) how the communication interacts with the Fetch API, which has well expected interactions with WebRequest API.  You could then (semi-redundantly) describe how extensions could interact with the new functionality in non-normative text.

So, I appreciate that there isn't a WebExtensions W3C spec, but adding new ways of pages communicating w/ hosts, w/o describing and setting expectations for how clients and extensions can mediate that communication is scary from a privacy perspective (since, for most people, extensions are the main way people claw back any control over their privacy)

**Comment by ricea** [comment]
In general, it *doesn't* interact with the Fetch API. However, we can normatively describe how it interacts with CSP.

**Comment by martinthomson** [comment]
So this is a really interesting problem.  On the one hand, Adam is right regarding the effect that this might have on the final specification that is built here.  I agree that it isn't necessary to document what affordances are being made for content filtering and blocking.

However, I think that it is important to recognize the effect that design choices have on these techniques.  Content filtering or blocking is very common for a variety of reasons.  A protocol that uses stable identifiers (like domain names) provides these tools good handles on which it can operate.  As the current design relies on URLs with domain names, it would seem that the needs of blockers are accounted for, with one exception.

Use of self-signed certificates, which can be authorized through the provision of a certificate fingerprint, makes it possible to avoid name-based blocking.  I don't particularly like that feature for a range of reasons, including this.  Though this is more of a problem for the IETF working group, which is responsible for concepts like resource identity, I believe that it is important to consider in the design.  My current view is that if self-signed certificates are that important (I'd prefer not to keep them, honestly), maybe the only way to deal with that is make them trivially distinguishable as such.

**Comment by yutakahirano** [comment]
Chrome is planning to update https://developer.chrome.com/docs/extensions/reference/webRequest/ for the WebTransport over HTTP/3 support.

**Comment by yutakahirano** [comment]
I think we don't need to "resolve" this issue by the initial release.

**Comment by jan-ivar** [comment]
> In general, it _doesn't_ interact with the Fetch API.

Not the API, but we've recently integrated WebTransport more closely with [fetch spec algorithms](https://w3c.github.io/webtransport/#initialize-webtransport-over-http) if that helps web extensions.

> Chrome is planning to update https://developer.chrome.com/docs/extensions/reference/webRequest/ for the WebTransport over HTTP/3 support.

@yutakahirano This sounds promising! Would you anticipate this resolving all issues in the OP?

@mixedpuppy @martinthomson Would this solution satisfy needs from a Mozilla pov, or do we have input?

**Comment by yutakahirano** [comment]
Both the API and protocol has changed much since this issue was filed. 

> If I'm mulitplexing over the existing HTTP3 connection, how can I block the Web Transport request (since no new request is happening)
> Most blocking tools focus on blocking the initial request. The spec seems to envision connections that require no additional request, just receiving server push messages (for example, layering over HTTP3 and just listening). How can I block these?

Our implementation allows extensions to block WebTransport over HTTP/3 session establishment. A WebTransport over HTTP/3 session is started with an HTTP CONNECT request and response, and the handshake is exposed to extensions similar to usual HTTP request/response. I expect that won't change even when we support `allowPooling`. Extensions cannot observe or intervene after that - it cannot know when a stream is created, for example.

> Can blocking tools see into the communication, and block some messages but not others?

No.

cc: @yoichio

**Comment by yutakahirano** [comment]
https://github.com/GoogleChrome/developer.chrome.com/pull/1128

**Comment by jan-ivar** [comment]
The [spec](https://w3c.github.io/webtransport/#initialize-webtransport-over-http) now uses fetch to connect using the [obtain a connection](https://fetch.spec.whatwg.org/#concept-connection-obtain) algorithm.

See also @yutakahirano's comment directly about Chrome supporting webtransport in the Web Request api.

@pes10k can this be closed now?

**Comment by pes10k** [comment]
Hi @jan-ivar ; that change would address the issue for me.  I don't see that change in the doc though.  Can you link me to the PR or part in the spec?  The only mention i see in the spec about fetch is about cert validation.  Am i missing something?

**Comment by jan-ivar** [comment]
@pes10k it's [here](https://w3c.github.io/webtransport/#initialize-webtransport-over-http): _"5. If [should request be blocked by Content Security Policy?](https://w3c.github.io/webappsec-csp/#should-block-request) with request returns "**Blocked**", or if request [should be blocked due to a bad port](https://fetch.spec.whatwg.org/#block-bad-port) returns **blocked**, then abort the remaining steps ..."_ — The latter is fetch's block-list algorithm.

**Comment by jan-ivar** [comment]
Based on the last comment this seems resolved. @pes10k please reopen if this is not the case.

**Comment by pes10k** [comment]
hi @jan-ivar .  Thanks for the tag in.  Im not sure i understand though; this text seems to be specific to content security policy no? Is there a convention or definition being used here that I'm missing (that CSP is short hand for "ways that a user agent might interpose on the page at the user's discretion" or similar)?  If so, I'd appreciate it if you could point me to other similar uses or definitions so i could catch up.

If not though, could you say. a bit more about how the seemingly-CSP-specific text addresses the concern?

**Comment by jan-ivar** [comment]
The [second part](https://w3c.github.io/webtransport/#initialize-webtransport-over-http) is not CSP-specific: _"or if request [should be blocked due to a bad port](https://fetch.spec.whatwg.org/#block-bad-port) returns blocked, then abort the remaining steps ..." — The latter is fetch's block-list algorithm."_

The text two steps below should also be relevant: _"Let connection be the result of [obtaining a connection](https://fetch.spec.whatwg.org/#concept-connection-obtain)"_ This shows WebTransport using the fetch spec to open new HTTP/3 or HTTP/2 connections or even using existing ones from the browser's connection pool.

This hopefully satisfies the criteria that communication interact with the Fetch API, which has well expected interactions with WebRequest API.

**Comment by pes10k** [comment]
Ah, apologies @jan-ivar i was following the links you posted in https://github.com/w3c/webtransport/issues/96#issuecomment-1030423689, since i didn't see any commits referencing this issue since then.

But the current text looks good to me. All good with closing this out. Thanks
